File: 11.PrefixSum前缀和/leetcode_3212_numberOfSubmatrices.py
from typing import List

class Solution:
    def numberOfSubmatrices(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        preSum = [[[0,0] for _ in range(n+1)] for _ in range(m+1)]
        res = 0
        for i in range(m):
            for j in range(n):
                preSum[i+1][j+1][0] = preSum[i+1][j][0]+preSum[i][j+1][0]-preSum[i][j][0]+(grid[i][j]=='X')
                preSum[i+1][j+1][1] = preSum[i+1][j][1]+preSum[i][j+1][1]-preSum[i][j][1]+(grid[i][j]=='Y')
                if preSum[i+1][j+1][0] == preSum[i+1][j+1][1] and preSum[i+1][j+1][0] != 0:
                    res += 1
        return res

if __name__ == "__main__":
    s = Solution()
    grid = [["X","Y","."],["Y",".","."]]
    print(s.numberOfSubmatrices(grid))

File: 11.PrefixSum前缀和/leetcode_3212_numberOfSubmatrices.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int numberOfSubmatrices(vector<vector<char>>& grid) {
        int m = grid.size(), n = grid[0].size();
        vector<vector<pair<int, int>>> mat(m+1, vector<pair<int, int>>(n+1, make_pair(0, 0)));
        int res = 0;
        for (int i=0; i<m; i++){
            for (int j=0; j<n; j++){
                mat[i+1][j+1].first = mat[i+1][j].first+mat[i][j+1].first-mat[i][j].first+(grid[i][j]=='X');
                mat[i+1][j+1].second = mat[i+1][j].second+mat[i][j+1].second-mat[i][j].second+(grid[i][j]=='Y');
                if (mat[i+1][j+1].first==mat[i+1][j+1].second && mat[i+1][j+1].first){
                    res++;
                }
            }
        }
        return res;
    }
};

int main(){
    Solution s;
    vector<vector<char>> grid = {{'X','Y','.'},{'Y','.','.'}};
    cout << s.numberOfSubmatrices(grid) << endl;
}
